java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -142,16 +138,19 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
         case LESS_THAN:
         case LESS_THAN_OR_EQUAL_TO:
           symbolicValues = programState.peekValues(2);
-          setAsUndetermined(symbolicValues.get(1));
-          setAsUndetermined(symbolicValues.get(0));
+          removeZeroConstraint(symbolicValues.get(1));
+          removeZeroConstraint(symbolicValues.get(0));
           break;
         default:
           // do nothing
       }
     }
 
-    private void setAsUndetermined(SymbolicValue sv) {
-      programState = programState.addConstraint(sv, new ZeroConstraint(DivByZeroStatus.UNDETERMINED));
+    private void removeZeroConstraint(SymbolicValue sv) {
+      PMap<Class<? extends Constraint>, Constraint> constraints = programState.getConstraints(sv);
+      if(constraints != null) {
+        programState = programState.addConstraints(sv, constraints.remove(ZeroConstraint.class));
```

Comment:
This looks definitely counterintuitive, as we add something with a remove on the other side. Logic, as we replace the map by a map without the constraint, but still requires to know perfectly well the data structure of the domains. 

Maybe a dedicated method in `ProgramState`? the very same construction is already present at multiple places.